src/common/transform.h

```diff
@@ -0,0 +1,211 @@
+        // Ignore other attributes of GPUDistribution for spliting index.
+        size_t shard_size =
+            GPUDistribution::Block(devices).ShardSize(range_size, d);
+        Range shard_range {0, static_cast<Range::DifferenceType>(shard_size)};
+        dh::safe_cuda(cudaSetDevice(d));
+        const int GRID_SIZE =
+            static_cast<int>(dh::DivRoundUp(*(range_.end()), kBlockThreads));
+
+        detail::LaunchCUDAKernel<<<GRID_SIZE, kBlockThreads>>>(
+            _func, shard_range, UnpackHDV(_vectors, d)...);
+        dh::safe_cuda(cudaGetLastError());
+        dh::safe_cuda(cudaDeviceSynchronize());
+      }
+    }
+#else
+    /*! \brief Dummy funtion defined when compiling for CPU.  */
+    template <typename std::enable_if<!CompiledWithCuda>::type* = nullptr,
+              typename... HDV>
+    void LaunchCUDA(Functor _func, HDV*... _vectors) const {
+      LOG(FATAL) << "Not part of device code. WITH_CUDA: " << WITH_CUDA();
+    }
+#endif
+
+    template <typename... HDV>
+    void LaunchCPU(Functor func, HDV*... vectors) const {
+      auto end = *(range_.end());
+#pragma omp parallel for schedule(static, 1)
+      for (omp_ulong idx = 0; idx < end; ++idx) {
+        func(idx, UnpackHDV(vectors)...);
+      }
+    }
+
+   private:
+    /*! \brief Callable object. */
+    Functor func_;
+    /*! \brief Range object specifying parallel threads index range. */
+    Range range_;
+    /*! \brief Whether resharding for vectors is required. */
+    bool reshard_;
+    GPUDistribution distribution_;
+  };
+
+ public:
+  /*!
+   * \brief Initialize a Transform object.
+   *
+   * \tparam Functor  A callable object type.
+   * \return A Evaluator having one method Eval.
+   *
+   * \param func    A callable object, accepting a size_t thread index,
+   *                  followed by a set of Span classes.
+   * \param range   Range object specifying parallel threads index range.
+   * \param devices GPUSet specifying GPUs to use, when compiling for CPU,
+   *                  this should be GPUSet::Empty().
+   * \param reshard Whether Reshard for HostDeviceVector is needed.
+   */
+  template <typename Functor>
+  static Evaluator<Functor> Init(Functor func, Range const range,
+                                 GPUSet const devices,
+                                 bool const reshard=true) {
+    return Evaluator<Functor> {func, range, devices, reshard};
+  }
+  template <typename Functor>
+  static Evaluator<Functor> Init(Functor func, Range const range,
+                                 GPUDistribution const dist,
+                                 bool const reshard=true) {
+    return Evaluator<Functor> {func, range, dist, reshard};
+  }
+};
+
+template <typename HDV>
+void ReshardAll(HDV* vector, GPUDistribution dist) {
+  vector->Reshard(dist);
+}
+template <typename HDV, typename... HdvDist>
+void ReshardAll(HDV* vector, GPUDistribution dist, HdvDist... rest) {
+  vector->Reshard(dist);
+  ReshardAll(rest...);
+}
```

Comment:
Please remove this (I mean `ReshardAll`).

Functions like these improve neither performance nor readability of the code. On the contrary, they increase complexity by introducing yet another unnecessary level of abstraction.

Just calling `Reshard()` on each HostDeviceVector sequentially is clearer.

Reply:
You are right.  These functions will be removed.